File: Algorithms/src/2130. Maximum Twin Sum of a Linked List/Code/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func pairSum(head *ListNode) int {

    // Find the middle of the linked list
    slow := head
    fast := head

    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }

    // Reverse the second half
    var prev *ListNode = nil

    for slow != nil {
        nextNode := slow.Next
        slow.Next = prev
        prev = slow
        slow = nextNode
    }

    // Calculate maximum twin sum
    ans := 0
    first := head
    second := prev

    for second != nil {
        sum := first.Val + second.Val

        if sum > ans {
            ans = sum
        }

        first = first.Next
        second = second.Next
    }

    return ans
}

File: Algorithms/src/2130. Maximum Twin Sum of a Linked List/Code/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {

        // Find the middle of the linked list
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse the second half
        ListNode prev = null;

        while (slow != null) {
            ListNode nextNode = slow.next;
            slow.next = prev;
            prev = slow;
            slow = nextNode;
        }

        // Calculate maximum twin sum
        int ans = 0;
        ListNode first = head;
        ListNode second = prev;

        while (second != null) {
            ans = Math.max(ans, first.val + second.val);

            first = first.next;
            second = second.next;
        }

        return ans;
    }
}